formats/json/commonMain/src/kotlinx/serialization/json/JsonNamingStrategy.kt

```diff
@@ -0,0 +1,107 @@
+    /**
+     * Accepts an original [serialName] (defined by property name in the class or [SerialName] annotation) and returns
+     * a transformed serial name which should be used for serialization and deserialization.
+     *
+     * Besides string manipulation operations, it is also possible to implement transformations that depend on the [descriptor]
+     * and its element (defined by [elementIndex]) currently being serialized.
+     * It is guaranteed that `descriptor.getElementName(elementIndex) == serialName`.
+     * For example, one can choose different transformations depending on [SerialInfo]
+     * annotations (see [SerialDescriptor.getElementAnnotations]) or element optionality (see [SerialDescriptor.isElementOptional]).
+     *
+     * Note that invocations of this function are cached for performance reasons.
+     * Caching strategy is an implementation detail and shouldn't be assumed as a part of the public API contract, as it may be changed in future releases.
+     * Therefore, it is essential for this function to be pure: it should not have any side effects, and it should
+     * return the same String for a given [descriptor], [elementIndex], and [serialName], regardless of the number of invocations.
+     */
+    public fun serialNameForJson(descriptor: SerialDescriptor, elementIndex: Int, serialName: String): String
+
+    /**
+     * Contains basic, ready to use naming strategies.
+     */
+    @ExperimentalSerializationApi
+    public companion object Builtins {
+
+        /**
+         * A strategy that transforms serial names from camel case to snake case — lowercase characters with words separated by underscores.
+         * The descriptor parameter is not used.
+         *
+         * It applies to every character following transformation rules:
+         *
+         * 1. If character `C` is in upper case, and the previous character exists, was not uppercase, and was not underscore, character `C` is transformed into underscore + c.lowercaseChar(): `_c`.
+         * 2. If character `C` is in upper case but does not match other conditions from 1., it is transformed into lowercase: `c`. Thus, upper case acronyms like URL are transformed correctly.
+         * 3. Otherwise, the character remains intact.
+         *
+         * **Note on cases:** Whether a character is in upper case is determined by the result of [Char.isUpperCase] function.
+         * Lowercase transformation is performed by [Char.lowercaseChar], not by [Char.lowercase],
+         * and therefore does not support one-to-many and many-to-one character mappings.
+         * See the documentation of these functions for details.
+         */
+        @ExperimentalSerializationApi
+        public val SnakeCase: JsonNamingStrategy = object : JsonNamingStrategy {
+            override fun serialNameForJson(descriptor: SerialDescriptor, elementIndex: Int, serialName: String): String =
+                buildString(serialName.length * 2) {
```

Comment:
In some way the `serialName` parameter is redundant. In addition, it is not clear whether this property refers to the serial name at the useSite (`serialDescriptor.getElementName(index)`) or the type (`serialDescriptor.getElementDescriptor(index).serialName`).

I would say that if the parameter is retained, the parameter should be named such to make this clear.

Reply:
@shanshin I'm a bit reluctant to provide such corner-case handling because it will make documentation for this state machine much harder to understand. Although it may be reduced to just several examples?

Reply:
@pdvrieze A whole strategy is about properties serial names, so I think it won't be confused with classes

Reply:
Maybe it's better to describe not how the string is transformed character by character, but to describe the rules where the underscore is inserted?
eg 
- between characters when a lowercase letter is followed by a capital letter
- between two capital letters, if after the second letter there is a lowercase
- between characters when a non-letter char is followed by a capital letter

the resulting string is converted to lowercase

Reply:
I've tried to do this in new docs, check them our